src/movement/maestro.hpp

```diff
@@ -0,0 +1,342 @@
+ * @author Ryan Summers
+ * @date 10-25-2016
+ *
+ * @brief Provides declaration and definitions of the thruster controller
+ *        class.
+ *
+ * @note Any documentation specified in this file refers to information located
+ *       at https://www.pololu.com/docs/0J40. Information is formatted into
+ *       section and subsection. For example, (4.5) refers to Section 4
+ *       subsection 5.
+ */
+
+#ifndef THRUSTER_CONTROLLER_H
+#define THRUSTER_CONTROLLER_H
+
+#include "utility/serial.hpp"
+#include <ros/ros.h>
+#include <vector>
+
+using std::vector;
+
+namespace rs {
+    /**
+     * Thruster controlling class.
+     *
+     * @brief This class provides an interface for communicating with the
+     *        Pololu Maestro servo controller. These servo commands are then
+     *        translated into a thruster power by the ESC (electronic speed
+     *        controller).
+     *
+     * @tparam thrusters The number of thrusters to be controlled by the class.
+     * @tparam max_speed The software limit on the maximum speed that any
+     *         thruster may be set to.
+     */
+    class ThrusterController
+    {
+        /*
+         * Specifies the delay in seconds after which a reset signal (1500
+         * microsecond pulse) should be sent.
+         */
+        static constexpr double reset_delay = 120.0;
+
+        public:
+            /**
+             * Constructor.
+             *
+             * @param _thrusters The number of thrusters controlled by the
+             *        controller.
+             * @param _max_speed The maximum normalized speed in the forward
+             *        and reverse directions that should be allowed by the
+             *        controller.
+             */
+            ThrusterController(const int _thrusters, const float _max_speed) :
+                        port(NULL),
+                        thrusters(_thrusters),
+                        max_speed(_max_speed)
+            {
+            }
+
+            /**
+             * Deconstructor.
+             *
+             * @brief This function sends zeroing commands to the maestro on
+             *        class destruction.
+             */
+            ~ThrusterController()
+            {
+                if (port != NULL)
+                {
+                    vector<float> zero;
+                    for (int i = 0; i < thrusters; ++i)
+                    {
+                        zero.push_back(0.0);
+                    }
+                    set(zero);
+                }
+            }
+
+            /**
+             * Sets the serial port used to communicate with the thrusters.
+             *
+             * @param _port A pointer to a configured and initialized serial
+             *        port to be used for communicating with the Maestro.
+             *
+             * @return Zero upon success and -1 upon failure.
+             */
+            int setPort(Serial *_port)
+            {
+                port = _port;
```

Comment:
I would say it would be fine if the argument name did not have an underscore, however, for most this would cause confusion as they are backwards from convention.